_posts/2019-03-05-build-firefox-extension-with-ringcentral-embeddable.md

```diff
@@ -0,0 +1,304 @@
+  );
+}
+```
+
+You can get full google client codes in here. Firstly, we register authorize path when we register service. Then we build a message listener to response authorize request.
+
+### Integrate Google contacts into widget
+
+Add following codes into background.js
+
+```js
+chrome.runtime.onMessage.addListener(function(request, sender, sendResponse) {
+  if (request.type === 'rc-register-service') {
+    registerService(sendResponse)
+    return true;
+  }
+  if (request.type === 'rc-post-message-request') {
+    if (request.path === '/authorize') {
+      onAuthorize(request.body.authorized);
+      sendResponse({ data: 'ok' });
+    }
+    if (request.path === '/contacts') {
+      onGetContacts(request, sendResponse);
+    }
+    return true;
+  }
+});
+
+async function registerService(sendResponse) {
+  const authorized = googleClient.checkAuthorize();
+  sendResponse({
+    action: 'registerService',
+    service: {
+      name: 'Google',
+      authorizationPath: '/authorize',
+      authorizedTitle: 'Unauthorize',
+      unauthorizedTitle: 'Authorize',
+      authorized,
+      contactsPath: '/contacts',
+    }
+  });
+}
+
+async function onGetContacts(request, sendResponse) {
+  const pageToken = request.body.page === 1 ? null : request.body.page;
+  const syncToken = request.body.syncTimestamp;
+  const response = await googleClient.queryContacts({ pageToken, syncToken });
+  const contacts = response.connections || [];
+  sendResponse({
+    data: contacts.map((c) => ({
+      id: c.resourceName.replace('people/', ''),
+      name: c.names[0] && c.names[0].displayName,
+      type: 'Google', // need to same as service name
+      phoneNumbers:
+        (c.phoneNumbers && c.phoneNumbers.map(p => ({ phoneNumber: p.value, phoneType: p.type }))) ||
+        [],
+      emails: (c.emailAddresses && c.emailAddresses.map(c => c.value)) || [],
+    })),
+    nextPage: response.nextPageToken,
+    syncTimestamp: response.nextSyncToken,
+  })
+}
+```
+
+Firstly, we register contact path when register service. Then we respond to contacts request.
+
+For more feature integration and details, you can get full codes in [here](https://github.com/ringcentral/ringcentral-embeddable-for-google-firefox-addon).
+
+We also provide CRM Firefox extension factory CLI tool to help developers build Firefox extension. You can get it [here](https://github.com/ringcentral/ringcentral-embeddable-extension-factory). Just run it and select Firefox extension, and you can get it work.
+
+## Try it out
+
+To get release packages [here](https://github.com/ringcentral/ringcentral-embeddable-for-google-firefox-addon/releases) and try it out by installing into Firefox. Reach out to us on [GitHub issues](https://github.com/ringcentral/ringcentral-embeddable-for-google-firefox-addon/issues), [Developer Community](https://devcommunity.ringcentral.com/ringcentraldev), or [Stack Overflow](https://stackoverflow.com/questions/tagged/ringcentral) if you have any questions.
```

Comment:
get release packages => get `the` release packages
by installing into => by installing `them` into